untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,461 @@
+It is also used extensively in programming language theory research and functional language design.
+
+There are two broad kinds of lambda calculus, _typed_, and _untyped_.
+The _untyped_ kind, discussed in this article, expresses an unrestricted computation.
+However, as a system of logic, it is self-contradictory.
+The _typed_ kind, which we intend to discuss in a follow-up article, is a restricted version.
+It doesn't have the logical paradox that plagues the untyped kind and can serve as a logic system.
+
+The list of languages based on lambda calculus includes:
+
+- Haskell
+- Most languages in the ML family, like OCaml and F#
+- Cardano blockchain's Plutus Core
+
+There's a rumor that John McCarthy based Lisp on lambda calculus, but this seems to be, broadly speaking, false.
+The original Lisp used different name resolution semantics, and as much was stated by John himself:
+
+> ... one of the myths concerning LISP that people think up or invent for themselves becomes apparent, and that is that LISP is somehow a realization of the lambda calculus, or that was the intention.
+> The truth is that I didn't understand the lambda calculus, really.
+>
+> -- John McCarthy, Lisp session, History of Programming Languages
+
+Also, while not strictly speaking based on lambda calculus, C++ template metalanguage is a functional language in its own right, so much of what we discuss below happens to be applicable.
+In fact, one could either embed lambda calculus into C++ template metalanguage or write a transpiler from a lambda calculus-based language into C++ template metalanguage.
+See, for instance, [Translating Lambda Calculus into C++ Templates by Vít Šefl](https://dx.doi.org/10.1007/978-3-030-83978-9_5).
+
+## A bit of history
+
+> 1936 - Alan Turing invents every programming language that will ever be but is shanghaied by British Intelligence to be 007 before he can patent them.
+>
+> 1936 - Alonzo Church also invents every language that will ever be but does it better.
+> His lambda calculus is ignored because it is insufficiently C-like.
+> This criticism occurs in spite of the fact that C has not yet been invented.
+>
+> -- James Iry, [A Brief, Incomplete, and Mostly Wrong History of Programming Languages][brief-incomplete-history]
+
+[brief-incomplete-history]: http://james-iry.blogspot.com/2009/05/brief-incomplete-and-mostly-wrong.html
+
+Lambda calculus, initially envisioned as a formal logic system, was developed by Alonzo Church around the 1930s to explore the foundations of mathematics.
+The initial formulation had a logical inconsistency known as [Kleene–Rosser paradox].
+To sidestep this issue, Church isolated the part of lambda calculus relevant only to computation in 1936.
+This isolate is now known as the untyped lambda calculus.
+Later, In 1940, a typed version of lambda calculus was introduced, which has weaker expressive power but is logically consistent.
```

Comment:
One may also mention that the idea of types has been initially proposed by Bertrand Russell (1903, 1907, 1911). The initial version of a typed lambda was a combination of the lambda calculus and Russell's type theory.

Reply:
I've added a mention that typed λ-calculus is based on Russel's type theory. Getting into it beyond that seems slightly off-topic. Hopefully I'll remember to get into more detail in the planned simply typed λ-calculus article.